### PR TITLE
[REVIEW] Add DeviceBuffer.release() cdef method

### PR DESCRIPTION
**Pull request by shwina**
This is to support releasing the contents of a `cudf.Column` to a `unique_ptr[column]`, something that has been requested by the Blazing team.

(How) do we test this?

cc: @kkraus14 @jakirkham

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by kkraus14** [comment]
For testing can we create a buffer and round trip it through something like `rmm.DeviceBuffer.c_from_unique_ptr(dbuf.c_release())`?

**Comment by jakirkham** [comment]
Could you please share a bit more about the use case?

File: python/rmm/_lib/device_buffer.pyx [hunk 1]
```diff
@@ -290,6 +290,13 @@ cdef class DeviceBuffer:
     cdef void* c_data(self):
         return self.c_obj.get()[0].data()
 
+    cdef unique_ptr[device_buffer] release(self):
```
Comment (kkraus14) [inline python/rmm/_lib/device_buffer.pyx:293]: All of the other cdef functions start with `c_`, may as well follow the trend here.

File: python/rmm/_lib/device_buffer.pyx [hunk 2]
```diff
@@ -290,6 +290,13 @@ cdef class DeviceBuffer:
...
+        """
+        Releases ownership the data held by this DeviceBuffer.
+        """
+        self.ptr = self.size = self.stream = 0
+        return move(self.c_obj)
```
Comment (jakirkham) [inline python/rmm/_lib/device_buffer.pyx:298]: This probably breaks other methods here. Thoughts on putting a trivial `device_buffer` in its place?

Comment (shwina) [reply python/rmm/_lib/device_buffer.pyx:298]: Can do!

**Comment by shwina** [comment]
@kkraus14 @jakirkham 

> Could you please share a bit more about the use case?

Blazing has asked for a cudf Cython function that passes ownership of the data/mask of a `Column` from Python back to C++. I don't know very much more about how they plan to use it but maybe Keith can shed some light. This is basically a pre-requisite for that.

> For testing can we create a buffer and round trip it through something like `rmm.DeviceBuffer.c_from_unique_ptr(dbuf.c_release())`?

OK but we'll have to set up some additional machinery for building/running Cython tests. Definitely worth doing IMO.

**Comment by kkraus14** [comment]
> Blazing has asked for a cudf Cython function that passes ownership of the data/mask of a `Column` from Python back to C++. I don't know very much more about how they plan to use it but maybe Keith can shed some light. This is basically a pre-requisite for that.

Maybe @felipeblazing can give some additional context here, but I believe the BlazingSQL engine manages its memory at the C++ layer, so they need a way to hand memory from Python back to C++ zero copy and have the C++ manage the lifetime of it.

> OK but we'll have to set up some additional machinery for building/running Cython tests. Definitely worth doing IMO.

Agreed, RMM is moving towards more and more Cython where this makes sense.

File: python/rmm/_lib/device_buffer.pyx [hunk 3]
```diff
@@ -278,18 +278,25 @@ cdef class DeviceBuffer:
...
 
-    cpdef size_t capacity(self):
+    cpdef size_t capacity(self) except *:
         return self.c_obj.get()[0].capacity()
 
-    cdef void* c_data(self):
+    cdef void* c_data(self) except *:
         return self.c_obj.get()[0].data()
 
+    cdef device_buffer c_release(self) except *:
```
Comment (jakirkham) [inline python/rmm/_lib/device_buffer.pyx:293]: I think this should be `except +` if we are trying to wrap C++ exceptions. `except *` is for checking whether C/C++ code (not Cython) raises an exception (like when using Python's C API directly).

```suggestion
    cdef device_buffer c_release(self) except +:
```

File: python/rmm/_lib/device_buffer.pyx [hunk 4]
```diff
@@ -278,18 +278,25 @@ cdef class DeviceBuffer:
...
+    cpdef void resize(self, size_t new_size) except *:
         self.c_obj.get()[0].resize(new_size)
 
-    cpdef size_t capacity(self):
+    cpdef size_t capacity(self) except *:
         return self.c_obj.get()[0].capacity()
 
-    cdef void* c_data(self):
+    cdef void* c_data(self) except *:
         return self.c_obj.get()[0].data()
```
Comment (jakirkham) [inline python/rmm/_lib/device_buffer.pyx:291]: These are [marked `noexcept` in C++]( https://github.com/rapidsai/rmm/blob/bb7ff451b9ce3b98e7c9bdc148ffea40880144c4/include/rmm/device_buffer.hpp#L330-L360 ). So I don't think we need these.

File: python/rmm/_lib/device_buffer.pyx (hunk 3 above)
Comment (jakirkham) [reply python/rmm/_lib/device_buffer.pyx:293]: Actually scratch that. The [`.release()` method is `noexcept`]( https://en.cppreference.com/w/cpp/memory/unique_ptr/release ). So this shouldn't be needed either.

File: python/rmm/_lib/device_buffer.pyx (hunk 4 above)
Comment (shwina) [reply python/rmm/_lib/device_buffer.pyx:291]: 👍

Comment (shwina) [reply python/rmm/_lib/device_buffer.pyx:291]: But wait -- we are doing more than just calling `self.c_obj.data()` for example. What if `c_obj.get()` throws, or the dereferencing. Those exceptions get swallowed.

File: python/rmm/_lib/device_buffer.pyx (hunk 3 above)
Comment (shwina) [reply python/rmm/_lib/device_buffer.pyx:293]: In `c_release()` we are also doing `self.ptr = 0`, which threw for me initially as `ptr` was a `property`. That exception got swallowed. In general, I think it's safest to `except *` anywhere that we are returning a non-Python object

File: python/rmm/_lib/device_buffer.pyx (hunk 4 above)
Comment (jakirkham) [reply python/rmm/_lib/device_buffer.pyx:291]: Well [`.get()` is also marked `noexcept`]( https://en.cppreference.com/w/cpp/memory/unique_ptr/get ) and `.data()` was already included in the `noexcept` above.

In any event, if there are C++ functions/methods that Cython exposes that do raise exceptions, these should be marked correctly in Cython (or fixed as needed) as opposed to handling them here.

File: python/rmm/_lib/device_buffer.pyx (hunk 3 above)
Comment (jakirkham) [reply python/rmm/_lib/device_buffer.pyx:293]: > In general, I think it's safest to except * anywhere that we are returning a non-Python object

Ah right. Not sure why I missed this. Yeah agreed.

File: python/rmm/_lib/device_buffer.pyx (hunk 4 above)
Comment (shwina) [reply python/rmm/_lib/device_buffer.pyx:291]: > In any event, if there are C++ functions/methods that Cython exposes that do raise exceptions, these should be marked correctly in Cython (or fixed as needed) as opposed to handling them here.

There are two kinds of `except` clauses to consider here: 

1. Exceptions thrown by C++: to correctly capture and re-raise these as Python `RuntimeError`s, we need to mark the wrapped functions as `except +`.

2. Python exceptions thrown in `cdef` functions: for `cdef` functions that don't return a Python object (like this one), any Python exceptions that may occur within them -- including the `RuntimeError`s mentioned in (1)  -- will be ignored, unless we mark those `cdef` functions `except *` or `except ?`.

(1) and (2) are somewhat independent, in that just doing one does not eliminate the need for the other. I agree that for C++ functions marked `noexcept`, we can safely omit (1).

But omitting (2) is dangerous and not future proof. It's too easy for an implementation of a `cdef` function to change to include calls to C++/Python code that can throw. Furthermore, even if all the code inside a `cdef` function is "guaranteed" not to throw, it's still possible for runtime errors to originate from inside that function, as I demonstrated here: https://github.com/pandas-dev/pandas/pull/33723

**Comment by shwina** [comment]
> Why are we having to catch if c_obj is a nullptr everywhere?

Currently, we segfault if `c_obj` is a nullptr. Thus calling `DeviceBuffer.size` after calling `release()` on it, for example, will segfault. We can keep that behaviour and call it UB. @jrhemstad any thoughts here?

**Comment by jrhemstad** [comment]
> > Why are we having to catch if c_obj is a nullptr everywhere?
> 
> Currently, we segfault if `c_obj` is a nullptr. Thus calling `DeviceBuffer.size` after calling `release()` on it, for example, will segfault. We can keep that behaviour and call it UB. @jrhemstad any thoughts here?

I'm not sure what the question is that I should be weighing in on.

File: python/rmm/_lib/device_buffer.pyx [hunk 5]
```diff
@@ -278,17 +278,35 @@ cdef class DeviceBuffer:
 
         return b
 
-    cdef size_t c_size(self):
-        return self.c_obj.get()[0].size()
+    cdef size_t c_size(self) except *:
+        if self.c_obj:
+            return self.c_obj.get()[0].size()
+        else:
+            return 0
```
Comment (shwina) [inline python/rmm/_lib/device_buffer.pyx:285]: After handing ownership of the data of a device buffer from Python back to C++, should it be UB/segfault to call any Python methods such as `DeviceBuffer.size`? Should we return 0/None? Or should we raise? @jakirkham @kkraus14 

In any case, would it be better to leave `c_size` as is, and handle the check for `self.c_obj` in the Python method `size` instead?

**Comment by shwina** [comment]
> I'm not sure what the question is that I should be weighing in on.

Yeah apologies - I guess this is actually a Cython/Python problem. 

Continuing discussion in https://github.com/rapidsai/rmm/pull/427/files#r447244315

File: python/rmm/_lib/device_buffer.pyx (hunk 5 above)
Comment (jrhemstad) [reply python/rmm/_lib/device_buffer.pyx:285]: fwiw, I think a "released-from" buffer should be similar to a "moved-from" buffer, i.e., it's the same as an empty buffer.

Comment (jrhemstad) [reply python/rmm/_lib/device_buffer.pyx:285]: In fact, if you just simply construct the returned `unique_ptr<device_buffer>` by move constructing from the wrapped `device_buffer`, then you shouldn't need to do anything special casing and you'll just rely on the behavior of a moved-from `device_buffer` being the same as an empty buffer.

File: python/rmm/_lib/device_buffer.pyx [hunk 6]
```diff
@@ -278,17 +278,35 @@ cdef class DeviceBuffer:
...
+            return <void*><uintptr_t>0
+
+    cdef device_buffer c_release(self) except *:
+        """
+        Releases ownership the data held by this DeviceBuffer.
+        """
+        if self.c_obj:
+            return cython.operator.dereference(self.c_obj.release())
+        else:
+            return device_buffer(0)
```
Comment (jakirkham) [inline python/rmm/_lib/device_buffer.pyx:309]: Could we assign and empty `device_buffer` to `c_obj` after we release the existing one? That should protect against errors in other cases.

File: python/rmm/_lib/device_buffer.pyx (hunk 5 above)
Comment (jakirkham) [reply python/rmm/_lib/device_buffer.pyx:285]: Having segfaults in Python sounds dangerous. That could come up anywhere and have unexpected affects.

I think we should just fill in `c_obj` to have an empty `device_buffer` after releasing. This should make sure other methods behave ok with sensible defaults.

Comment (shwina) [reply python/rmm/_lib/device_buffer.pyx:285]: Thanks @jrhemstad - that worked!

File: python/rmm/_lib/device_buffer.pyx (hunk 6 above)
Comment (shwina) [reply python/rmm/_lib/device_buffer.pyx:309]: So we now call `move()` on the device buffer underlying `c_obj` as Jake suggested, and that has the same effect as assigning a unique_ptr to an empty device_buffer to `c_obj`  :)

File: python/rmm/tests/test_device_buffer.pyx [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    got = DeviceBuffer.c_from_unique_ptr(
+        make_unique[device_buffer](buf.c_release())
+    )
+    np.testing.assert_equal(expect.copy_to_host(), got.copy_to_host())
+
+
+def test_size_after_release():
+    cdef DeviceBuffer buf = DeviceBuffer.to_device(b'abc')
+    buf.c_release()
+    print(buf.size)
```
Comment (jakirkham) [inline python/rmm/tests/test_device_buffer.pyx:40]: Should this `print` be dropped?

```suggestion
```

File: python/rmm/tests/test_device_buffer.pyx [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import numpy as np
```
[the remark this answers is not in the document]
Comment (jakirkham) [reply python/rmm/tests/test_device_buffer.pyx:21]: Looks like this was done above.

File: python/rmm/tests/test_device_buffer.pyx (hunk 7 above)
Comment (jakirkham) [reply python/rmm/tests/test_device_buffer.pyx:40]: Or should we be doing something with the value like `assert`ing?

File: python/rmm/_lib/device_buffer.pyx (hunk 6 above)
Comment (jakirkham) [reply python/rmm/_lib/device_buffer.pyx]: Ah ok. Didn't realize `move` would ensure this. Thought that was UB.

File: python/rmm/tests/test_device_buffer.pyx [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def test_size_after_release():
+    cdef DeviceBuffer buf = DeviceBuffer.to_device(b'abc')
+    buf.c_release()
+    print(buf.size)
+
+
+test_release()
+test_size_after_release()
```
Comment (kkraus14) [inline python/rmm/tests/test_device_buffer.pyx:44]: Are these leftover from debugging?

File: python/rmm/_lib/device_buffer.pyx (hunk 6 above)
Comment (jrhemstad) [reply python/rmm/_lib/device_buffer.pyx]: Note that `move` by itself does _not_ guarantee this. It's 100% defined by the class's move constructor about what is the state of a "moved from" object. 

Here's what `std::vector` says: 

> Constructs the container with the contents of other using move semantics. Allocator is obtained by move-construction from the allocator belonging to other. After the move, other is guaranteed to be empty().

Here's what `device_buffer` says: https://github.com/rapidsai/rmm/blob/branch-0.15/include/rmm/device_buffer.hpp#L155

File: python/rmm/tests/test_device_buffer.pyx (hunk 9 above)
Comment (shwina) [reply python/rmm/tests/test_device_buffer.pyx:44]: No, this is how we run Cython tests at the moment :P

Comment (shwina) [reply python/rmm/tests/test_device_buffer.pyx:44]: Trying to figure out something better here.

File: python/setup.py [hunk 10]
```diff
@@ -30,24 +31,59 @@
...
+    "../include",
+    "../build/include",
+    os.path.dirname(sysconfig.get_path("include")),
+    cuda_include_dir,
 ]
 
+library_dirs = [get_python_lib(), os.path.join(os.sys.prefix, "lib")]
+
+# lib:
+extensions = cythonize(
```
Comment (kkraus14) [inline python/setup.py:50]: Looks like you're still calling `cythonize` on `extensions` in the actual `setup` call, I think that needs to be removed?

Comment (shwina) [reply python/setup.py:50]: Good catch - fixed.

File: python/rmm/tests/test_device_buffer.pyx (hunk 7 above)
Comment (shwina) [reply python/rmm/tests/test_device_buffer.pyx]: Fixed :)

**Pull request merged**